src/plugins/workspace/server/permission_control/client.ts

```diff
@@ -30,6 +33,8 @@ export class SavedObjectsPermissionControl {
   private readonly logger: Logger;
   private _getScopedClient?: SavedObjectsServiceStart['getScopedClient'];
   private auth?: HttpAuth;
+  private _savedObjectCache: WeakMap<LegacyRequest, { [key: string]: SavedObject }> = new WeakMap();
```

Comment:
What about we making it `WeakMap<string(request.id), { [key: string]: SavedObject }>`? By doing so we do not need to import ensureRawRequest and LegacyRequest from core.

Reply:
We need to use Map instead of WeakMap if we want to use `request.id` as map key. Key of WeakMap should be an object.

Reply:
Then what about using Map? I just do not think it is necessary to import the internal `ensureRawRequest`.

Reply:
In my point of view, `WeakMap` has better memory performance on large requests. There are many request level caches are using `WeakMap`. 
Here are the examples: 
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/61d5c4905bce1709d29629ade132723d87fb6897/src/core/server/http/base_path_service.ts#L41
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/61d5c4905bce1709d29629ade132723d87fb6897/src/core/server/http/auth_headers_storage.ts#L46
Some [external plugins](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/61d5c4905bce1709d29629ade132723d87fb6897/src/plugins/console/server/routes/api/console/proxy/create_handler.ts#L38) are using `ensureRawRequest` too. I'm open to change to `Map` and using `request.uuid` as the map key. What do you think about it?

Reply:
Update with `request.uuid` instead of `request.id`, the `request.id` is not guaranteed to be unique per request. See below comments:
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/http/router/request.ts#L165